packages/mui-joy/src/Switch/Switch.tsx

```diff
@@ -0,0 +1,257 @@
+import { SwitchProps } from './SwitchProps';
+
+const useUtilityClasses = (ownerState: SwitchProps & { focusVisible: boolean }) => {
+  const { classes, checked, disabled, focusVisible, readOnly } = ownerState;
+
+  const slots = {
+    root: [
+      'root',
+      checked && 'checked',
+      disabled && 'disabled',
+      focusVisible && 'focusVisible',
+      readOnly && 'readOnly',
+    ],
+    thumb: ['thumb', checked && 'checked'],
+    track: ['track', checked && 'checked'],
+    input: ['input'],
+  };
+
+  return composeClasses(slots, getSwitchUtilityClass, classes);
+};
+
+const SwitchRoot = styled('span', {
+  name: 'MuiSwitch',
+  slot: 'Root',
+  overridesResolver: (props, styles) => styles.root,
+})<{ ownerState: SwitchProps }>(({ theme, ownerState }) => {
+  return {
+    '--Switch-track-radius': theme.vars.radius.lg,
+    '--Switch-track-width': '48px',
+    '--Switch-track-height': '24px',
+    '--Switch-thumb-size': '16px',
+    ...(ownerState.size === 'sm' && {
+      '--Switch-track-width': '40px',
+      '--Switch-track-height': '20px',
+      '--Switch-thumb-size': '12px',
+    }),
+    ...(ownerState.size === 'lg' && {
+      '--Switch-track-width': '64px',
+      '--Switch-track-height': '32px',
+      '--Switch-thumb-size': '24px',
+    }),
+    '--Switch-thumb-radius': 'calc(var(--Switch-track-radius) - 2px)',
+    '--Switch-thumb-width': 'var(--Switch-thumb-size)',
+    '--Switch-thumb-offset':
+      'max((var(--Switch-track-height) - var(--Switch-thumb-size)) / 2, 0px)',
+    display: 'inline-block',
+    width: 'var(--Switch-track-width)', // should have the same width as track because flex parent can stretch SwitchRoot.
+    borderRadius: 'var(--Switch-track-radius)',
+    position: 'relative',
+    padding:
+      'calc((var(--Switch-thumb-size) / 2) - (var(--Switch-track-height) / 2)) calc(-1 * var(--Switch-thumb-offset))',
+    color: theme.vars.palette.neutral.containedBg,
+    '&:hover': {
+      color: theme.vars.palette.neutral.containedBg,
+    },
+    [`&.${switchClasses.checked}`]: {
+      color: theme.vars.palette[ownerState.color!].containedBg,
+      '&:hover': {
+        color: theme.vars.palette[ownerState.color!].containedHoverBg,
+      },
+    },
+    [`&.${switchClasses.disabled}`]: {
+      pointerEvents: 'none',
+      cursor: 'default',
+      opacity: 0.6,
+    },
+    [`&.${switchClasses.focusVisible}`]: theme.focus.default,
+  };
+});
+
+const SwitchInput = styled('input', {
+  name: 'MuiSwitch',
+  slot: 'Input',
+  overridesResolver: (props, styles) => styles.input,
+})<{ ownerState: SwitchProps }>(() => ({
+  margin: 0,
+  height: '100%',
+  width: '100%',
+  opacity: 0,
+  position: 'absolute',
+  top: 0,
+  left: 0,
+  bottom: 0,
+  right: 0,
+  cursor: 'pointer',
+}));
+
+const SwitchTrack = styled('span', {
+  name: 'MuiSwitch',
+  slot: 'Track',
+  overridesResolver: (props, styles) => styles.track,
+})<{ ownerState: SwitchProps & { focusVisible: boolean } }>(() => ({
```

Comment:
This reminds me I need to export the SwitchOwnerState interface to avoid construction of such types ad hoc.